### PR TITLE
no conversion from T to not_null<T>

### PR DESCRIPTION
**Pull request by akrzemi1**

### CODE REVIEW

File: tests/notnull_tests.cpp [hunk 1]
```diff
@@ -80,20 +80,20 @@ SUITE(NotNullTests)
...
-        not_null<Unrelated*> t = reinterpret_cast<Unrelated*>(p.get());
+        not_null<Unrelated*> t(reinterpret_cast<Unrelated*>(p.get()));
         CHECK((void*)p.get() == (void*)t.get());
     }
 
     TEST(TestNotNullAssignment)
     {
         int i = 12;
-        not_null<int*> p = &i; 
+        not_null<int*> p(&i);
```
Comment (rianquinn) [inline tests/notnull_tests.cpp]: If we are getting rid of the assignment operator of T, do we need the test at all? I believe there are already tests for construction.

Comment (gdr-at-ms) [inline tests/notnull_tests.cpp]: That line isn't exercising assignment though.

Comment (viboes) [inline tests/notnull_tests.cpp]: Do we need a make_not_null factory ?
```
void f(not_null<int*>);
f(make_not_null(&i))
```

Wondering if in the particular case of `not_null<T*>` it couldn't be copy-constructed and assigned from a `T&`.

```
f(i); // implicit conversion from T& to `not_null<T*>`
```

Comment (huangqinjin) [inline tests/notnull_tests.cpp]: @neilmacintosh `make_not_null()` is not a good idea, but what about `gsl::addressof()` just like the one in `std`?

**Comment by annagrin** [comment]
please see discussion in #699. Closiing this since we decided to remove the explicit constructor and add strict_not_null type for practical purposes.

**Pull request closed without merging**